### PR TITLE
auth/kubernetes: support for dynamically reloading short-lived tokens

### LINKED ISSUE #12855: Kubernetes Auth Method documentation not compatible with Kubernetes v1.21+
## Describe the bug

Kubernetes v1.21 introduced new bound service account tokens with following new properties:

1. Tokens are bound to a pod and they get revoked when pod is deleted.
2. Tokens will expire.
3. Tokens are audience-limited.

The Kubernetes Auth Method documentation does not cover the use of the new tokens. This results in failed authentication.

## To Reproduce

Pre-condition:

Deploy Vault inside Kubernetes. Configure Kubernetes Auth Method according to examples at https://www.vaultproject.io/docs/auth/kubernetes. Set JWT value to `token_reviewer_jwt="$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"`.

**(1) Authentication fails because of invalidated token**

1. Delete the pod that runs Vault and wait for it to be recreated.
2. Try to authenticate using Kubernetes Auth Method
3. Vault returns `403 Forbidden` with `{ "errors": [ "permission denied" ] }` in response body.

Vault prints following error:

```
2021-10-18T12:09:01.075Z [ERROR] auth.kubernetes.auth_kubernetes_726dbb64: login unauthorized due to: lookup failed: service account unauthorized; this could mean it has been deleted or recreated with a new token
```

Kubernetes API server prints following error

```
E1018 12:09:01.074341       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, Token has been invalidated]"
```

**(2) Authentication fails because of expired token**

1. Configure Kubernetes cluster with a short expiration time by setting API server parameter `service-account-max-token-expiration` to `1h`. Wait for 1 hour for the token to expire .
2. Try to authenticate using Kubernetes Auth Method. 
3. Vault returns `403 Forbidden` with `{ "errors": [ "permission denied" ] }` in response body.

Vault prints following error:

```
2021-10-18T11:47:23.756Z [ERROR] auth.kubernetes.auth_kubernetes_97f02f17: login unauthorized due to: lookup failed: service account unauthorized; this could mean it has been deleted or recreated with a new token
```

Kubernetes API server prints following error

```
E1018 11:47:23.755730       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, Token has expired.]"
```

## Proposed fix

Include following PR
* https://github.com/hashicorp/vault-plugin-auth-kubernetes/pull/122

Update documentation, taking following points into account:

* It is recommended NOT to store Kubernetes service account token persistently into config when running Vault inside Kubernetes. Instead, the recommendation is to use the local token and CA files. 
* When running Vault outside of Kubernetes, user needs to use legacy token which is not bound to any pod and does not expire. User should be instructed how to access legacy token via Secret directly. User should be WARNED that the legacy tokens could be removed in later Kubernetes release completely.

I also think following point would be beneficial

* Recommend `disable_iss_validation=True`. The benefit is that user does not need to find the correct `iss` from their specific cluster (or change it if it changes later). Kubernetes Token Review API will validate the token, so I do not see negative side in recommending that.

### PR DESCRIPTION
**Pull request by tsaarni**
This change is continuation to https://github.com/hashicorp/vault-plugin-auth-kubernetes/pull/122 where support was added for dynamically reloading Kubernetes short-lived tokens for better Kubernetes 1.21+ compatibility.  This change takes the feature into use in Vault.

* Uplift new version of Kubernetes auth plugin that does not store the  service account token persistently to Vault storage.
* Update the documentation to recommend local token again when running  Vault inside cluster.

Fixes #12855

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -96,7 +96,7 @@ require (
 	github.com/hashicorp/vault-plugin-auth-gcp v0.11.3
 	github.com/hashicorp/vault-plugin-auth-jwt v0.11.4
 	github.com/hashicorp/vault-plugin-auth-kerberos v0.5.0
-	github.com/hashicorp/vault-plugin-auth-kubernetes v0.11.3
+	github.com/hashicorp/vault-plugin-auth-kubernetes v0.7.1-0.20220107030939-d289258274b7
```
Comment (tsaarni) [inline go.mod:99]: @tvoran Would it be possible to tag a new release of vault-plugin-auth-kubernetes so that we can rever to proper version number?

**Comment by tsaarni** [comment]
@tomhjp, @tvoran I've sent this PR to uplift hashicorp/vault-plugin-auth-kubernetes#122 into Vault. 

I added proposal for documentation changes, though I feel they are maybe bit unnecessarily complicated for in-cluster deployment. User should not need to be aware about short-lived tokens anymore as reloading is transparent - though I guess the information is necessary since user can be running older version of Vault.

(for running Vault outside-cluster, I'm not sure if copying service account tokens from cluster has ever been that good idea - I think client cert support might have been better)

File: changelog/13595.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+```release-note:improvement
+auth/kubernetes: Added support for short-lived tokens for Kubernetes 1.21+ compatibility
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply changelog/13595.txt]: Fixed!

**Review by tomhjp — COMMENTED** [review]
Thanks for raising this! A few small suggestions. This should definitely make it into 1.10. For now we should probably just get this merged into main, and then it will automatically go live on the website when 1.10 gets released but not before.

File: website/content/docs/auth/kubernetes.mdx [hunk 3]
```diff
@@ -168,6 +168,23 @@ short-lived tokens. If you would like to disable this, set
 
 [k8s-extended-tokens]: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/#options
 
+#### Use local service account token as the reviewer JWT
+
+When running Vault in a Kubernetes pod the recommended option is to use local
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply website/content/docs/auth/kubernetes.mdx]: Fixed!

File: website/content/docs/auth/kubernetes.mdx [hunk 4]
```diff
@@ -168,6 +168,23 @@ short-lived tokens. If you would like to disable this, set
...
+service account token. Vault will periodically re-read the file to support
+short-lived tokens. Local token and CA certificate are used by configuring the
+auth method without setting `token_reviewer_jwt` and `kubernetes_ca_cert`.
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply website/content/docs/auth/kubernetes.mdx]: This is good clarification, thanks! Taken it as such :)

File: website/content/docs/auth/kubernetes.mdx [hunk 5]
```diff
@@ -168,6 +168,23 @@ short-lived tokens. If you would like to disable this, set
...
+
+```bash
+vault write auth/kubernetes/config \
+    kubernetes_host=https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_PORT
+```
+
+!> **Note:** Requires Vault VERSION_TODO. In earlier versions the service account
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply website/content/docs/auth/kubernetes.mdx]: Added!

**Review by tvoran — APPROVED** [review]

**Pull request merged**